### PR TITLE
`info` and `eldoc` ops: accept a Compliment-style `context` parameter

### PR DESCRIPTION
**Pull request by vemv**
Fixes https://github.com/clojure-emacs/cider-nrepl/issues/812

Tested out locally - works as intended!

Cheers - V

### CODE REVIEW

File: src/cider/nrepl/middleware/info.clj [hunk 1]
```diff
@@ -68,10 +105,9 @@
                              {:env env
                               :dialect :cljs}))]
     (cond
+      java? (info/info-java class (or member sym))
       (and ns sym) (info/info* info-params)
-      (and class member) (info/info-java class member)
-      :else (throw (Exception.
```
Comment (vemv) [inline src/cider/nrepl/middleware/info.clj:73]: The `throw` wasn't compatible with the new logic. nil is more flexible and will result in a graceful op-specific reply (e.g. status `no-info` or `no-eldoc`)

**Comment by vemv** [comment]
@alexander-yakushev - feel free to check out this PR. You may be happy to see Compliment used in a new way!

Maybe later as things get more time-proven, we may want to have an explicit API just for this kind of purpose.

File: src/cider/nrepl/middleware/info.clj [hunk 2]
```diff
@@ -57,9 +59,44 @@
           blacklist
           util/transform-value))))
 
+(defn extract-class-from-compliment [ns-str context]
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/info.clj:62]: I docstring will be nice here, especially given this is a public function. (I know it's clear what it does, but still)

File: src/cider/nrepl/middleware/info.clj [hunk 3]
```diff
@@ -57,9 +59,44 @@
...
+        ;; We can always be analyzing a broken context.
+        nil))))
+
 (defn info
-  [{:keys [ns sym symbol class member] :as msg}]
-  (let [[ns sym class member] (map misc/as-sym [ns (or sym symbol) class member])
+  [{:keys [ns sym class member context]
+    legacy-sym :symbol
+    :as msg}]
+  (let [sym (or (not-empty legacy-sym)
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/info.clj:78]: Why do you need `not-empty` here?

File: src/cider/nrepl/middleware/info.clj [hunk 4]
```diff
@@ -57,9 +59,44 @@
...
+                (not-empty sym))
+        class (try
+                (or (when (and (seq class)
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/info.clj:81]: Is this nil-punning here? I still think it doesn't read very well. :D

**Comment by bbatsov** [comment]
Other than the slightly confusing usage of `not-empty` and `seq` here and there it seems OK overall.

File: src/cider/nrepl/middleware/info.clj (hunk 3 above)
Comment (vemv) [reply src/cider/nrepl/middleware/info.clj:81]: For discarding empty strings

I don't necessarily expect them, but I've found it's a good habit to handle them.

File: src/cider/nrepl/middleware/info.clj (hunk 4 above)
Comment (vemv) [reply src/cider/nrepl/middleware/info.clj:84]: My understanding is that you'd prefer...?

```clj
(when (and (seq class) (seq ns))
  (find-ns (symbol ns)))
```

can do!

Comment (vemv) [reply src/cider/nrepl/middleware/info.clj:84]: Not sure of what the alternative would be

(I had written a suggestion, but it wasn't accurate)

Comment (bbatsov) [reply src/cider/nrepl/middleware/info.clj:84]: I'm fine either way - I just don't like much mixing the use of `seq` and `not-empty`. Rich liked `seq`, but `not-empty` was other later on for those who felt `seq` wasn't very intention revealing.

Comment (bbatsov) [reply src/cider/nrepl/middleware/info.clj:84]: (my main point was the one I made earlier - that probably checking for empty strings is redundant, although I understand your reasoning as well)

File: src/cider/nrepl/middleware/info.clj (hunk 3 above)
Comment (vemv) [reply src/cider/nrepl/middleware/info.clj:81]: 🤝

In an ideal world we'd have something like Spec checking at the edges

File: src/cider/nrepl/middleware/info.clj (hunk 4 above)
Comment (vemv) [reply src/cider/nrepl/middleware/info.clj:84]: A nice convention is to exclusively use `seq` as a condition, and `not-empty` as an object one intends to further use

**Pull request merged**